projects/output-and-error-handling/main.go

```diff
@@ -0,0 +1,96 @@
+package main
+
+import (
+	"bufio"
+	"fmt"
+	"net/http"
+	"os"
+	"time"
+)
+
+func sendRequest(url string) *http.Response {
+	resp, err := http.Get(url)
+	if err != nil {
+		fmt.Println("We can't get the weather!")
+		fmt.Fprintln(os.Stderr, "Failed to initialize.!", err)
+		os.Exit(2)
+	}
+	if resp == nil {
+		fmt.Fprintln(os.Stderr, "The server did not reply anything, which here is considered an error!")
+		os.Exit(52)
+	}
+	return resp
+}
+
+func handleStatusCode(resp *http.Response, sleepTime *time.Duration) {
+	defer resp.Body.Close()
+	if resp.StatusCode == 200 {
+		scanner := bufio.NewScanner(resp.Body)
+		for i := 0; scanner.Scan(); i++ {
+			fmt.Println(scanner.Text())
+		}
+		if err := scanner.Err(); err != nil {
+			panic(err)
+		}
+		os.Exit(0)
+	} else if resp.StatusCode == 429 {
+		// If I receive 429 but can't find ["Retry-After"] so I determine 1s to sleep. It act same as a while
+		retryAfter := resp.Header.Get("Retry-After")
+		if retryAfter == "" || retryAfter == "a while" {
+			// I added with the previous sleep times to prevent sleeping for too long and more than 5s give up.
+			*sleepTime += time.Second
+		} else {
+
+			*sleepTime = parseRetryAfterNumber(retryAfter)
+		}
+		if *sleepTime > 5*time.Second {
+			fmt.Println("We can't get the weather!")
+			fmt.Fprintln(os.Stderr, "Sleep time is more than 5 second so HTTP page not retrieved. Returned another error with the HTTP error code being 400 or above.")
+			os.Exit(22)
+		} else if *sleepTime > time.Second {
+			fmt.Println("Things may be a bit slow because we're doing a retry!")
+		}
+		time.Sleep(*sleepTime)
+	} else {
+		fmt.Fprintln(os.Stderr, "Failed to initialize!")
+		os.Exit(2)
+	}
+}
+
+func parseRetryAfterNumber(retryAfter string) time.Duration {
+	if sleepTime, err := time.ParseDuration(retryAfter + "s"); err != nil {
+		return parseRetryAfterString(retryAfter)
+	} else {
+		return sleepTime
+	}
+}
+
+func parseRetryAfterString(retryAfter string) time.Duration {
+	sleepTime := time.Second
+	if timeStamp, err := time.Parse(http.TimeFormat, retryAfter); err == nil {
+		sleepTime = time.Until(timeStamp)
+	} else {
+		fmt.Fprintln(os.Stderr, "There was an error in the value of Retry-After", err)
+	}
+	return sleepTime
+}
+
+func main() {
+	var sleepTime time.Duration
+	for {
```

Comment:
If the server keeps replying with a 429 and, say, a Retry-After header saying to retry in 1 second, what will this code do? Is that the most desirable behaviour?

Reply:
It will stop after 5sec. That's why I sent the pointer and added with last values of sleepTimes.
```
*sleepTime += time.Second
```